test/unit/range/container/container_test.cpp

```diff
@@ -0,0 +1,362 @@
+
+    // n * value
+    TypeParam t3;
+    t3.assign(2, dna4::C);
+    EXPECT_EQ(t3, t0);
+
+    // from another container's sub-range
+    TypeParam t4;
+    t4.assign(t1.cbegin(), t1.cend());
+    EXPECT_EQ(t4, t1);
+
+    // initializer list
+    TypeParam t5, t6;
+    t5.assign({dna4::A, dna4::C, dna4::C, dna4::G, dna4::T});
+    t6 = {dna4::A, dna4::C, dna4::C, dna4::G, dna4::T};
+    EXPECT_EQ(t5, t1);
+    EXPECT_EQ(t6, t1);
+
+    // direct from another container
+    if constexpr (!std::is_same_v<TypeParam, std::vector<dna4>>)
+    {
+        TypeParam t7;
+        t7.assign("ACCGT"_dna4);
+        EXPECT_EQ(t7, t1);
+    }
+}
+
+TYPED_TEST(container, iterators)
+{
+    TypeParam t1{dna4::A, dna4::C, dna4::C, dna4::G, dna4::T};
+    TypeParam const t2{dna4::A, dna4::C, dna4::C, dna4::G, dna4::T};
+
+    // begin
+    EXPECT_EQ(*t1.begin(),  dna4::A);
+    EXPECT_EQ(*t1.cbegin(), dna4::A);
+    EXPECT_EQ(*t2.begin(),  dna4::A);
+    EXPECT_EQ(*t2.cbegin(), dna4::A);
+
+    // end and arithmetic
+    EXPECT_EQ(*(t1.end()  - 1), dna4::T);
+    EXPECT_EQ(*(t1.cend() - 1), dna4::T);
+    EXPECT_EQ(*(t2.end()  - 1), dna4::T);
+    EXPECT_EQ(*(t2.cend() - 1), dna4::T);
+
+    // convertibility between const and non-const
+    EXPECT_TRUE(t1.cend() == t1.end());
+
+    // mutability
+    *t1.begin() = dna4::T;
+    EXPECT_TRUE((ranges::equal(t1, "TCCGT"_dna4)));
+}
+
+TYPED_TEST(container, element_access)
+{
+    TypeParam t1{dna4::A, dna4::C, dna4::C, dna4::G, dna4::T};
+    TypeParam const t2{dna4::A, dna4::C, dna4::C, dna4::G, dna4::T};
+
+    // at
+    EXPECT_EQ(t1.at(0), dna4::A);
+    EXPECT_EQ(t2.at(0), dna4::A);
+
+    //TODO check at's ability to throw
+
+    // []
+    EXPECT_EQ(t1[0], dna4::A);
+    EXPECT_EQ(t2[0], dna4::A);
+
+    // front
+    EXPECT_EQ(t1.front(), dna4::A);
+    EXPECT_EQ(t2.front(), dna4::A);
+
+    // back
+    EXPECT_EQ(t1.back(), dna4::T);
+    EXPECT_EQ(t2.back(), dna4::T);
+
+    // mutability
+    t1[0] = dna4::T;
+    EXPECT_TRUE((ranges::equal(t1, "TCCGT"_dna4)));
+
+    t1.front() = dna4::C;
+    EXPECT_TRUE((ranges::equal(t1, "CCCGT"_dna4)));
+
+    t1.back() = dna4::G;
+    EXPECT_TRUE((ranges::equal(t1, "CCCGG"_dna4)));
+}
+
+TYPED_TEST(container, capacity)
+{
+    TypeParam t0{};
+    TypeParam t1{dna4::A, dna4::C, dna4::C, dna4::G, dna4::T};
+    TypeParam const t2{dna4::A, dna4::C, dna4::C, dna4::G, dna4::T};
+
+    // empty
+    EXPECT_TRUE(t0.empty());
+    EXPECT_FALSE(t1.empty());
+    EXPECT_FALSE(t2.empty());
+
+    // size
+    EXPECT_EQ(t0.size(), 0u);
+    EXPECT_EQ(t1.size(), 5u);
+    EXPECT_EQ(t2.size(), 5u);
+
+    // max_size
+    EXPECT_GT(t0.max_size(), 1'000'000'000'000u);
+    EXPECT_GT(t1.max_size(), 1'000'000'000'000u);
+    EXPECT_GT(t2.max_size(), 1'000'000'000'000u);
+
+    // capacity
+    EXPECT_GE(t0.capacity(), t0.size());
+    EXPECT_GE(t1.capacity(), t1.size());
+    EXPECT_GE(t2.capacity(), t2.size());
+
+    // reserve
+    EXPECT_LT(t0.capacity(), 1000u);
```

Comment:
I think we should only test functionality that is specified in the documentation. This isn't.